Download test data

keggsmurph21 opened the pull request:
@jonorthwash @mr-martian @BartholomaeusAstrum @calebho 

This branch implements downloading some test data from behind protected S3 endpoints so that `pytest` can consume it.  In order to ensure that this data is not publicly readable, the S3 bucket is configured so that only users with the correct AWS credentials can read from it (no one has write permissions).  For more background on S3 permissions management, see [the docs](https://docs.aws.amazon.com/AmazonS3/latest/dev/security-best-practices.html).

The permissions for the build system (CircleCI) are encoded as environment variables directly in the CircleCI app (see [the docs](https://circleci.com/docs/2.0/env-vars/)), which is only accessibly by administrators of the [swatphonlab](https://github.com/swatphonlab) org.

To be clear: individual users (i.e., people who aren't contributing to the project) will never have to download this test data -- it is strictly for `pytest`.

However, **if you are a developer of this project, you'll need to get these credentials in order to run our automated tests**.  Specifically, you'll need to set these environment variables: `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY`.

 Personally, I find writing the credentials to a file called `.env` in the repository root and then `source`ing that file to be useful.  That is:
```bash
$ cat .env
export AWS_ACCESS_KEY_ID=?????????????????????
export AWS_SECRET_ACCESS_KEY=?????????????????????????????????????????
$ source .env
```

**For a copy of these credentials, please reach out to me personally** (unless anyone has a better idea of how I might share them).

As a sanity check, `ls`ing the bucket + path with those environment variables set yields
```bash
$ source .env
$ aws s3 ls s3://swatphonlab/ultratrace-test-data/
2019-12-19 19:55:08          0 
2019-12-19 21:51:32  328112118 ftyers.tar.gz
2019-12-19 21:56:36 1488914073 qumuq.tar.gz
```
But
```bash
$ export AWS_ACCESS_KEY_ID=
$ export AWS_SECRET_ACCESS_KEY=
$ aws s3 ls s3://swatphonlab/ultratrace-test-data/

An error occurred (InvalidAccessKeyId) when calling the ListObjects operation: The AWS Access Key Id you provided does not exist in our records.
```

keggsmurph21 committed:
download-test-data: First draft to confirm env vars are working

keggsmurph21 committed:
circleci: Add a download-test-data run step

keggsmurph21 committed:
circleci: Pip-install boto3 so that we can download test data

keggsmurph21 committed:
download-test-data: Read S3 objects into GzipFiles

keggsmurph21 committed:
download-test-data: S3 should be a Client not a Resource!

keggsmurph21 committed:
download-test-data: Use S3Client::download_file directly

keggsmurph21 committed:
download-test-data: Automatically extract tarballs

This script also makes the download script much more user-friendly.
For example it prints out a long message in case all of the env vars
are not set.

Further, those env vars get set by the `.env` file, which is kept out
of source control, since it contains AWS secrets. For a copy of that
file, you can email [me](keggsmurph21 at gmail dot com) :^)

Some TODOs from this patch:
 - Fix the hashing/cacheing
 - Set up some more simple examples for testing

keggsmurph21 committed:
download-test-data: Fix `tar` syntax error

keggsmurph21 committed:
download-test-data: Implement hashing directory contents

This patch makes the script idempotent!  Also, it saves *lots* of
redundant network traffic.

calebho commented on scripts/download-test-data:
```diff
@@ -0,0 +1,127 @@
+# test data from the S3 tarballs.
+#
+# This value should be updated every time we make updates to ./test-data.
+MAGIC_HASH="326e33e80882799e088c0f34fb39aa8f"
```
Instead of hardcoding this here, why not just store it in S3 as well? Maybe we should have something like
* `latest`
* `v1`
  * `hash`
  * ... TAR files
* ...
* `vN`
  * `hash`
  * ... TAR files

where `latest` is a symlink to the latest version of `test-data`.

keggsmurph21 replied:
Good idea.  S3 doesn't support symlinks directly, but uploading a file containing only a string pointing to the next URL does the trick :^)

keggsmurph21 replied:
So I think this might be a bit more complicated than we originally thought.

Basically the issue is that under ^this scheme, we'd be storing a bunch of redundant data.  (Note: I'm making the assumption that our larger datasets won't change very frequently, especially as compared to our smaller / contrived ones.)  Consider the following hypothetical before / after `./test-data` directory tree:

**Before** (packaged as `v1` on S3):
```
./test-data/
    |- dataset_01/
    |   |- huge_file_01.tar.gz    [10gb]
    |- dataset_02/
        |- small_file_01.txt      [10kb]
```

**After** (added one `10kb` file):
```
./test-data/
    |- dataset_01/
    |   |- huge_file_01.tar.gz    [10gb]
    |- dataset_02/
        |- small_file_01.txt      [10kb]
        |- small_file_02.txt      [10kb]
```

Then if we were to package `./test-data` as `v2`, adding `10kb` of data forces us to store a redundant `10gb`.

Instead, I think we should push the version-control-ish scheme down one level -- to the actual datasets themselves.  That is, **Before** might look like this in S3:
```
s3://swatphonlab/ultratrace-test-data/
    |- manifest.json
    |- dataset_01/
    |   |- v1.tar.gz              [~10gb]
    |- dataset_02/
        |- v1.tar.gz              [~10kb]
```
with `manifest.json` containing
```json
{
    "datasets": {
        "dataset_01": {
            "hash": "xxxxxxxxx",
            "url": "s3://swatphonlab/ultratrace-test-data/dataset_01/v1.tar.gz"
        },
        "dataset_02": {
            "hash": "yyyyyyyyy",
            "url": "s3://swatphonlab/ultratrace-test-data/dataset_01/v1.tar.gz"
        }
    }
}
```
And **After**:
```
s3://swatphonlab/ultratrace-test-data/
    |- manifest.json
    |- dataset_01/
    |   |- v1.tar.gz              [~10gb]
    |- dataset_02/
        |- v1.tar.gz              [~10kb]
        |- v2.tar.gz              [~20kb]
```
with `manifest.json` containing
```json
{
    "datasets": {
        "dataset_01": {
            "hash": "xxxxxxxxx",
            "url": "s3://swatphonlab/ultratrace-test-data/dataset_01/v1.tar.gz"
        },
        "dataset_02": {
            "hash": "zzzzzzzzz",
            "url": "s3://swatphonlab/ultratrace-test-data/dataset_01/v2.tar.gz"
        }
    }
}
```
Here, adding `10kb` of data only costs us `~20kb` of cloud storage.  Also, the consumer could just check the individual datasets to see if the S3 hash has changed from whatever local hash (if any) they have.

Thoughts?

keggsmurph21 replied:
Actually, now that I think about it, we can probably just do
```
s3://swatphonlab/ultratrace-test-data/
    |- manifest.json
    |- dataset_01/
    |   |- v1.tar.gz              [~10gb]
    |- dataset_02/
        |- v1.tar.gz              [~10kb]
        |- v2.tar.gz              [~20kb]
```
with manifest.json containing
```json
{
    "datasets": {
        "dataset_01": "v1",
        "dataset_02": "v2"
    }
}
```
And then each local copy could keep a `version.txt` file (containing, e.g., `"v1"`) in the dataset directories and only update if they don't match the `manifest.json`.  Basically, I don't think we should spend too much time optimizing for versioned datasets ....

keggsmurph21 replied:
Implemented https://github.com/SwatPhonLab/UltraTrace/pull/116#discussion_r360642397 in e35e360 and 593a56a.

calebho reviewed (commented):
Not a bash expert, but overall lgtm

keggsmurph21 commented on .gitignore:
```diff
@@ -15,3 +16,6 @@ __pycache__
 .coverage
+
+# local environment setup
+.env
```
If you do end up putting your credentials in a file in this repo, make sure you never commit that file!

keggsmurph21 committed:
download-test-data: Split `hash-path` into its own script

Since we'll be moving to a symlink-ish system for storing test data
blobs, we'll want to have consistent hashing across different scripts.

Also, this allows us to be a bit more exacting in our attention to OS
interoperability.  Specifically, it fixes a bunch of issues that were
causing MacOS and Ubuntu to hash the same directory trees to different
values (and thus triggering excessive downloads).

keggsmurph21 committed:
FileLoaderBase: remove the `path = property(...)` line

This was causing issues for the class hierarchy: child classes were
not able to be instantiated because they were getting the error `Can't
instantiate abstract class with abstract methods: "path"`.  To work
around this, we can just require all references to `FileLoaderBase#path`
to be accessed via `::[gs]et_path()`.

keggsmurph21 committed:
ImageSetFileLoader: add abstract methods `::get_(height|width)`

keggsmurph21 committed:
DICOMLoader: Suppress "truncated" image errors

This is not a perfect fix, since now truncated images will just
silently be all black (after the truncation bit is encountered).

keggsmurph21 committed:
DICOMLoader: refactor `::from_file` to only throw `FileLoadError`s

keggsmurph21 committed:
DICOMLoader: Add tests (#104)

keggsmurph21 committed:
upload-test-data: (partial commit)

keggsmurph21 committed:
download-test-data: Refactor to Pythonworld

The management of the `manifest.json` is too complex to comfortably
handle in Bashworld (sadly).

keggsmurph21 committed:
upload-test-data: Add complement of download-"-"

This file will handle uploading new copies of old datasets and new
datasets.  For more context, see discussion on #116.

keggsmurph21 committed:
setup.py: Fix typo s/mypy-boto3/mypy-boto3-s3/

keggsmurph21 committed:
circleci: Install mypy-boto3-s3 (& merge commit)

keggsmurph21 committed:
download-test-data: Fix bug RE downloading to nonexistent folder

keggsmurph21 committed:
{up,down}load-test-data: Add better user-facing error messages

calebho commented on .circleci/config.yml:
```diff
@@ -6,6 +6,6 @@ jobs:
       - checkout
       - run: ./install.sh
-      - run: sudo pip install boto3 nox
+      - run: sudo pip install boto3 mypy_boto3_s3 nox
```
I don't think we need this extra dependency because the types shouldn't have any runtime effect

keggsmurph21 replied:
I'm not sure how we can get around importing `S3Client` though?  Because
```python
def download_file(s3: S3Client, ...) -> None: ...
```
throws a `NameError` otherwise ...

I could just make `s3: Any` to avoid this dependency.

calebho replied:
Nvm missed the import at the top

calebho commented on scripts/download-test-data:
```diff
@@ -0,0 +1,120 @@
+#!/usr/bin/env python
```
To be safe maybe do
```suggestion
#!/usr/bin/env python3
```

keggsmurph21 replied:
619b7e3

calebho commented on scripts/download-test-data:
```diff
@@ -0,0 +1,120 @@
+
+def hash_path(path: str) -> str:
+    assert os.path.exists(path)
+    proc = subprocess.run(
```
https://docs.python.org/3/library/subprocess.html#subprocess.check_output

keggsmurph21 replied:
619b7e3

calebho commented on scripts/download-test-data:
```diff
@@ -0,0 +1,120 @@
+PREFIX = "ultratrace-test-data"
+
+
+def usage() -> None:
```
Technically this is a `NoReturn`: https://docs.python.org/3/library/typing.html#typing.NoReturn
```suggestion
def usage() -> NoReturn:
```

keggsmurph21 replied:
619b7e3

calebho commented on scripts/upload-test-data:
```diff
@@ -0,0 +1,149 @@
+#!/usr/bin/env python
```
```suggestion
#!/usr/bin/env python3
```

keggsmurph21 replied:
619b7e3

calebho commented on scripts/upload-test-data:
```diff
@@ -0,0 +1,149 @@
+PREFIX = "ultratrace-test-data"
+
+
+def usage() -> None:
```
```suggestion
def usage() -> NoReturn:
```

keggsmurph21 replied:
619b7e3

keggsmurph21 committed:
{up,down}load-test-data: Fixes from PR comments
